ext/pdo/pdo_dbh.c

```diff
@@ -1382,9 +1376,7 @@ static void dbh_free(pdo_dbh_t *dbh, bool free_persistent)
 		pefree((char *)dbh->persistent_id, dbh->is_persistent);
 	}
 
-	if (!Z_ISUNDEF(dbh->def_stmt_ctor_args)) {
-		zval_ptr_dtor(&dbh->def_stmt_ctor_args);
-	}
+	/* The GC, via dbh_get_gc(), will free dbh->def_stmt_ctor_args() */
```

Comment:
[IMHO]

It might be better to free `dbh->def_stmt_ctor_args()` here as well.

Code like the following will cause a memory leak because `dbh_get_gc` is not called:
```
<?php
$db = new PDO('mysql:host=mysql;dbname=test;', 'root', ''); // my env
class Foo extends PDOStatement {}
$db->setAttribute(PDO::ATTR_STATEMENT_CLASS, array('Foo', ['param1']));
```

gdb:
```
(gdb) b dbh_get_gc
Breakpoint 1 at 0x364ad9: file /php-src/ext/pdo/pdo_dbh.c, line 1312.
(gdb) run /var/www/html/test_12154.php
Starting program: /php-src/sapi/cli/php /var/www/html/test_12154.php
[Mon Sep 11 15:25:29 2023]  Script:  '/var/www/html/test_12154.php'
/php-src/Zend/zend_hash.c(291) :  Freeing 0x00007ffff7a02600 (56 bytes), script=/var/www/html/test_12154.php
[Mon Sep 11 15:25:29 2023]  Script:  '/var/www/html/test_12154.php'
/php-src/Zend/zend_string.h(176) :  Freeing 0x00007ffff7a045c0 (32 bytes), script=/var/www/html/test_12154.php
[Mon Sep 11 15:25:29 2023]  Script:  '/var/www/html/test_12154.php'
/php-src/Zend/zend_hash.c(157) :  Freeing 0x00007ffff7a5e180 (136 bytes), script=/var/www/html/test_12154.php
=== Total 3 memory leaks detected ===
[Inferior 1 (process 4956) exited normally]
```

Reply:
Right, I was wondering about this. I'll add some tests for a GC and the non-GC case which made me think this (an SQLite bug)